cmd/ui.go

```diff
@@ -162,58 +202,100 @@ func showProgress(ctx context.Context, conf Config, execScheduler *local.Executi
 	var leftLen int64
 	for _, pb := range pbs {
 		l := pb.Left()
 		leftLen = lib.Max(int64(len(l)), leftLen)
 	}
-
 	// Limit to maximum left text length
-	leftMax := int(lib.Min(leftLen, maxLeftLength))
+	maxLeft := int(lib.Min(leftLen, maxLeftLength))
+
+	var progressBarsLastRender []byte
 
-	// For flicker-free progressbars!
-	progressBarsLastRender := []byte(renderMultipleBars(stdoutTTY, true, leftMax, pbs))
-	progressBarsPrint := func() {
+	printProgressBars := func() {
 		_, _ = stdout.Writer.Write(progressBarsLastRender)
 	}
 
+	var widthDelta int
+	// Default to responsive progress bars when in an interactive terminal
+	renderProgressBars := func(goBack bool) {
+		barText, longestLine := renderMultipleBars(stdoutTTY, goBack, maxLeft, termWidth, widthDelta, pbs)
+		widthDelta = termWidth - longestLine - termPadding
+		progressBarsLastRender = []byte(barText)
+	}
+
+	// Otherwise fallback to fixed compact progress bars
+	if !stdoutTTY {
+		widthDelta = -pb.DefaultWidth
+		renderProgressBars = func(goBack bool) {
+			barText, _ := renderMultipleBars(stdoutTTY, goBack, maxLeft, termWidth, widthDelta, pbs)
+			progressBarsLastRender = []byte(barText)
+		}
+	}
+
 	//TODO: make configurable?
 	updateFreq := 1 * time.Second
 	//TODO: remove !noColor after we fix how we handle colors (see the related
 	//description in the TODO message in cmd/root.go)
 	if stdoutTTY && !noColor {
 		updateFreq = 100 * time.Millisecond
 		outMutex.Lock()
-		stdout.PersistentText = progressBarsPrint
-		stderr.PersistentText = progressBarsPrint
+		stdout.PersistentText = printProgressBars
+		stderr.PersistentText = printProgressBars
 		outMutex.Unlock()
 		defer func() {
 			outMutex.Lock()
 			stdout.PersistentText = nil
 			stderr.PersistentText = nil
-			if ctx.Err() != nil {
-				// Render a last plain-text progressbar in an error
-				progressBarsLastRender = []byte(renderMultipleBars(stdoutTTY, false, leftMax, pbs))
-				progressBarsPrint()
-			}
 			outMutex.Unlock()
 		}()
 	}
 
+	var (
+		fd          = int(os.Stdout.Fd())
+		ticker      = time.NewTicker(updateFreq)
+		winchSignal = getWinchSignal()
+	)
+
+	// Default ticker-based progress bar resizing
+	updateTermWidth := func() {
+		<-ticker.C
+		termWidth, _, _ = terminal.GetSize(fd)
+	}
+	// More responsive progress bar resizing on platforms with SIGWINCH (*nix)
+	if winchSignal != nil {
+		winch := make(chan os.Signal, 1)
+		signal.Notify(winch, winchSignal)
+		updateTermWidth = func() {
+			select {
+			case <-ticker.C:
+			case <-winch:
+				termWidth, _, _ = terminal.GetSize(fd)
+			}
+		}
+	}
+
 	ctxDone := ctx.Done()
-	ticker := time.NewTicker(updateFreq)
 	for {
 		select {
-		case <-ticker.C:
-			barText := renderMultipleBars(stdoutTTY, true, leftMax, pbs)
-			outMutex.Lock()
-			progressBarsLastRender = []byte(barText)
-			progressBarsPrint()
-			outMutex.Unlock()
 		case <-ctxDone:
+			renderProgressBars(false)
+			printProgressBars()
 			return
+		default:
 		}
+		updateTermWidth()
```

Comment:
A minor issue, and maybe also what @MStoykov referenced above, but this `select` seems strange. I think it shouldn't have a default, but only 3 normal `case`s:
1. context is done, in which case it does what it currently does and exits
2. received a signal from the ticker, render progress bars
3. received a `winch` signal from, render progress bars

not a big deal right now with the 1s/100ms refresh rate, but if we allow setting custom refresh dates, this may hold up the script after it should be aborted, because if the context is interrupted while you wait on the timer, you're not going to abort immediately.

Reply:
Hm ... might've confused how this should work and made things worse ... but yeah I also find it strange.

 IMO updateTermWidth should go ... and the code should be :
```golang
var winch chan os.Signal
if sig := getWinchSignal(); sig !=nil {
 	winch = make(chan os.Signal, 1)
	signal.Notify(winch, sig)
}
for {
  select {
  case<-cxtDone:
  //what is currently + maybe locking outMutex ?
  case <-winch: // it is okay if it is nil in the case of the 
     termWidth, _, _ = terminal.GetSize(fd) 
  case <-ticker.C:
    if winch == nil {
 	  termWidth, _, _ = terminal.GetSize(fd) 
    }
  }
  // the rendering things
}
```
my whole previous thought process (if I remember correctly) was that we don't want to at all do anything on ticker.C if winch is defined which is not true :man_facepalming:  even if the size isn't changing we still want to update the progressbars, we just don't want to get the size if there is no need :D

Reply:
Thanks guys, it does make more sense to use a single `select`. I was trying to avoid checks in the `for` loop, but this is simpler. I'm not sure if we need to lock `outMutex`, but I suppose it's safer.

See 886859a6.

Reply:
LGTM now. `outMutex` would still need to be locked, otherwise something else (like the logger) could write something in the middle of the progressbars, no?